extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -553,6 +553,51 @@ export class Bind {
     return value;
   }
 
+  /**
+   * Returns a copy of the global state for a given field-based expression.
+   * It will wait for relevant amp-state components to finish fetching before returning.
+   *
+   * e.g. "foo.bar".
+   * @param {string} expr
+   * @return {!Promise<*>}
+   */
+  getStateWithWait(expr) {
+    const asyncLoadingAmpStates = {};
+    const ampStateEls = this.ampdoc.getRootNode().querySelectorAll('AMP-STATE');
+    let hasLoadingAmpState = false;
+    const gatherAsyncAmpStates = Promise.all(
+      toArray(ampStateEls).map(el => {
+        return whenUpgradedToCustomElement(el)
+          .then(() => el.getImpl(/* waitForBuild */ false))
+          .then(impl => {
+            const id = impl.element.getAttribute('id');
+            const loadingPromise = impl.getFetchAndUpdatePromise();
+            if (loadingPromise) {
+              asyncLoadingAmpStates[id] = loadingPromise;
+              hasLoadingAmpState = true;
+            }
+          });
+      })
+    );
+
+    return gatherAsyncAmpStates.then(() => {
+      if (!hasLoadingAmpState) {
+        return this.getState(expr);
+      }
+
+      let wait;
+      if (expr === '.') {
+        // If getting everything, then wait for all async amp states.
+        wait = Promise.all(Object.values(asyncLoadingAmpStates));
+      } else {
+        const stateKey = expr.split('.')[0];
+        wait = Promise.resolve(asyncLoadingAmpStates[stateKey]);
```

Comment:
How should we handle edge cases?  E.g. `asyncLoadingAmpStates[stateKey]` doesn't exist.

Reply:
Promise.resolve(undefined) actually does what I want for that case (immediately resolve). Are there other edge cases?

I can rewrite this clause to be more clear though:
```js
wait = stateKey in asyncLoadingAmpStates 
  ? asyncLoadingAmpStates[stateKey] 
  : Promise.resolve();
```

Reply:
Yea I mean from a user's POV. Is silent continuance what we want?

Reply:
Lets talk offline about this since I'm not sure I'm following. I don't think `getStateAsync` would be the right place for a warning or a failure. Bind expressions pointing to absent state continue silently, and I planned on following that pattern. 

If we did want to add warnings though, I'd imagine we'd want them to be handled by `amp-list` as opposed to `bind-impl`.

Reply:
Done.